pkg/cache/informer_cache.go

```diff
@@ -93,7 +93,11 @@ func (ip *informerCache) List(ctx context.Context, out runtime.Object, opts ...c
 		}
 		// http://knowyourmeme.com/memes/this-is-fine
 		elemType := reflect.Indirect(reflect.ValueOf(itemsPtr)).Type().Elem()
```

Comment:
For posterity, can we also add a comment here explaining what we're trying to do and why? I checked out the code and understood we're pretty much trying to figure out the type of the list passed in, but that might be a lot to ask for folks not familiar with the codebase.

Let's also move the logic to a local function and add some unit tests for it.

Reply:
Done, PTAL. I had to move the `UnconventionalListType` into a different package, because its now used by both unit and integration tests and would otherwise have caused an import cycle.